### DISCUSSION TITLE
How to debug client API

### DISCUSSION THREAD

**Discussion by ericm301** [dgreif/ring#1388] [Q&A]
Maybe I'm just a noob, and it's something simple, but I can't figure it out.

I got the rest-client-example to work in VSCode, and have modified it to get snapshots from one of the cameras. I want to get regular snapshots, and stitch them together as a timelapse. There is a function  `RingCamera.getPeriodicalFootage` that seems it would be useful; but, for some reason, I can't seem to get the response to retrieve anything.

I have tried debugging the client, but the .ts code is missing from the module. That code is in the client, but not the one in the node_modules folder. It just has the js and map files. I have tried compiling the ts, but I get crazy errors when I do.

How do I debug into the typescript files and not into the compiled JS?

**Comment by tsightler** [comment]
I mean, while your question is valid, there's not much debugging to do here on this specific function as it's just a wrapper for calling the URL to download such footage from Ring.  You can just make the request directly if you like using the request function from RestClient and substituting appropriate values for the variables:

https://api.ring.com/recordings/public/footages/${this.id}?start_at_ms=${startAtMs}&end_at_ms=${endAtMs}&kinds=online_periodical&kinds=offline_periodical

The call to that URL with proper values either returns something (an MP4 file), or it returns an error.  Basically, this is the same URL used by the Ring app when you open the timeline view.  It depends on having a Ring subscription and periodic snapshots enabled on the camera, otherwise there won't be anything to download from Ring.

**Reply by ericm301** [reply]
Here are the interfaces that are returned from getPeriodical...:

```
export interface PeriodicalFootage {
  start_ms: number
  end_ms: number
  playback_ms: number
  kind: 'online_periodical' | 'offline_periodical'
  url: string
  deleted: boolean
  snapshots: number[]
}
export interface PeriodicFootageResponse {
  meta: {
    pagination_key: number
    butch_size: number
  }
  data: PeriodicalFootage[]
  responseTimestamp: number
}
```

When I call getSnapshot I just get a Buffer that I save as a jpg. There's no explanation as to what these fields do.
The other (meta) issue is that I can't build the module without getting 'symbol used before init' errors, and the complied code is different than what's in the install package. So, I'm still confused. There must be build settings that I haven't configured properly.

**Reply by tsightler** [reply]
>There's no explanation as to what these fields do.

Welcome to using an API based on reverse engineering, we probably don't have a whole lot of idea what they do either, but you just poke around at it, most fields will be self-explanatory.

If you look at the actual code for the function:

```
getPeriodicalFootage(
    { startAtMs, endAtMs } = {
      startAtMs: getStartOfToday(),
      endAtMs: getEndOfToday(),
    },
  ) {
    // These will be mp4 clips that are created using periodic snapshots
    return this.restClient.request<PeriodicFootageResponse>({
      url: `https://api.ring.com/recordings/public/footages/${this.id}?start_at_ms=${startAtMs}&end_at_ms=${endAtMs}&kinds=online_periodical&kinds=offline_periodical`,
    })
  }
```

So you can call this function with no parameters at all and it will just return all of the periodical footage for that camera from the start of that day until the end.  The response will be in the format as described in those exports.  I've never even used this function before, but I just called it on a camera and the response for most fields was quite clear, the URL is the S3 URL to download the periodical footage, the start and end ms are the millisecond based epoc representing the start and end time of the snapshots in the given footage.  For me, I have all of my cameras set to snapshot every 30 seconds, which creates videos that play back for 60 seconds showing 120 snapshots covering a given hour.  Each hour is an element in the array returned.  You can then choose to download or stream the video from the URL in question.  If you pass a startAtMs or endAtMs epoch to the function it will return only any periodical footage available between those defined timestamps.

If you don't get anything back, then you either don't have a Ring subscription or the cameras aren't configured to take regular snapshots in the Ring app.
